Cofree comonad

edmundnoble opened the pull request:
Basic Cofree Comonad functionality, including a couple of instances. I tried to implement a few other things (`mapBranching(F ~> G)(Cofree[F, A]): Cofree[G, A]`, `decompile(F ~> W)(W[A]): Cofree[F, A]`) but I couldn't make them stack-safe. This implementation is inspired by the one in Scalaz.

adelbertc commented on free/src/main/scala/cats/free/Cofree.scala:
```diff
@@ -0,0 +1,124 @@
+  * A tree with data at the branches, as opposed to Free which is a tree with data at the leaves.
+  * Not an instruction set functor made into a program monad as in Free, but an instruction set's outputs as a
+  * functor made into a tree of the possible worlds reachable using the instruction set.
+  */
```
Was this derived/based on Scalaz's [Cofree](https://github.com/scalaz/scalaz/blob/series/7.3.x/core/src/main/scala/scalaz/Cofree.scala) or from scratch? If former could you add attribution to Scalaz, similar to what we have [here](https://github.com/typelevel/cats/blob/master/free/src/main/scala/cats/free/FreeT.scala#L14) ?

edmundnoble replied:
Totally derived from the one at Scalaz. I did not know there was an example of such an attribution somewhere in the code to crib from, I'll add it in.

adelbertc commented on free/src/test/scala/cats/free/CofreeTests.scala:
```diff
@@ -0,0 +1,118 @@
+
+  checkAll("Cofree[Option, ?]", ComonadTests[Cofree[Option, ?]].comonad[Int, Int, Int])
+  checkAll("Cofree[Option, ?]", FoldableTests[Cofree[Option, ?]].foldable[Int, Int])
+  checkAll("Cofree[Option, ?]", TraverseTests[Cofree[Option, ?]].traverse[Int, Int, Int, Int, Option, Option])
```
I think Traverse checks Foldable already

adelbertc replied:
Check Serializability of the Traverse instance?

edmundnoble replied:
I don't need to check that the Traverse and Foldable instances are valid separately? They are defined separately. I will add that serializable check.

adelbertc replied:
Are they? Where did you see that? Seem to be together https://github.com/typelevel/cats/blob/master/laws/src/main/scala/cats/laws/discipline/TraverseTests.scala#L9

adelbertc commented on free/src/test/scala/cats/free/CofreeTests.scala:
```diff
@@ -0,0 +1,118 @@
+  checkAll("Comonad[Cofree[Option, ?]]", SerializableTests.serializable(Comonad[Cofree[Option, ?]]))
+  checkAll("Cofree[List, ?]", ComonadTests[Cofree[List, ?]].comonad[Int, Int, Int])
+  checkAll("Cofree[List, ?]", FoldableTests[Cofree[List, ?]].foldable[Int, Int])
+  checkAll("Cofree[List, ?]", TraverseTests[Cofree[List, ?]].traverse[Int, Int, Int, Int, Option, Option])
```
Do we need to test these given the tests above? Seems the only difference is s/Option/List

edmundnoble replied:
If we're comfy with not testing these, absolutely. I don't see any precedent in other tests in doing the same law checks on different instances, so I'll go ahead and remove.

edmundnoble commented:
Is this alright for inclusion or should it perhaps have more operators?

adelbertc commented:
Could you write some functions for the other functions, maybe like `unfold` and `unfoldStrategy` ?

Also looks like the `Foldable`-specific instance isn't being tested because the test is getting the `Foldable` through the `Traverse` test. You can see an example [here](https://github.com/typelevel/cats/blob/master/tests/src/test/scala/cats/tests/NestedTests.scala#L107) of how we test these kinds of "if `F[_]: TC` then `MT[F, ?]: TC` law checks.

Other than that, LGTM 👍 Thanks @edmundnoble !

xuwei-k commented on free/src/main/scala/cats/free/Cofree.scala:
```diff
@@ -0,0 +1,128 @@
+}
+
+sealed abstract class CofreeInstances2 {
+  /** low priority `Foldable1` instance */
```
This is a `Foldable`, not `Foldable1`. There is `scalaz.Foldable1`. but cats does not have `Foldable1`. 

https://github.com/typelevel/cats/issues/145

adelbertc replied:
@xuwei-k We have `Foldable1` but we call it `Reducible` https://github.com/typelevel/cats/pull/298. No `Traverse1` though like you said below.

xuwei-k commented on free/src/main/scala/cats/free/Cofree.scala:
```diff
@@ -0,0 +1,128 @@
+}
+
+sealed abstract class CofreeInstances1 extends CofreeInstances2 {
+  /** low priority `Traverse1` instance */
```
This is a `Traverse`, not `Traverse1`

dwhitney commented:
Bump. This looks interesting and it looks like @edmundnoble has addressed @adelbertc's concerns. Any reason not to merge?

johnynek commented on free/src/main/scala/cats/free/Cofree.scala:
```diff
@@ -0,0 +1,135 @@
+
+  /** Evaluate just the tail. */
+  def forceTail: Cofree[S, A] =
+    Cofree[S, A](head, Eval.now(tailEval.value))
```
this is not tested, can we add a test?

johnynek commented on free/src/main/scala/cats/free/Cofree.scala:
```diff
@@ -0,0 +1,135 @@
+
+  /** Evaluate the entire Cofree tree. */
+  def forceAll(implicit S: Functor[S]): Cofree[S, A] =
+    Cofree[S, A](head, Eval.now(tailEval.map(S.map(_)(_.forceAll)).value))
```
this is not tested, can we add a test?

johnynek commented on free/src/main/scala/cats/free/Cofree.scala:
```diff
@@ -0,0 +1,135 @@
+  override final def reduceLeftTo[A, B](fa: Cofree[F, A])(z: A => B)(f: (B, A) => B): B =
+    F.foldLeft(fa.tailForced, z(fa.head))((b, cof) => foldLeft(cof, b)(f))
+
+  override def reduceRightTo[A, B](fa: Cofree[F, A])(z: A => B)(f: (A, Eval[B]) => Eval[B]): Eval[B] = {
```
why did you do this vs something like:

```scala
f(ta.head, fa.tailEval.map { reduceRightTo(_)(z)(f) })
```

edmundnoble replied:
Try doing that yourself. Doesn't typecheck as far as I can see; you're expecting `tailEval` to hold an `Eval[Cofree[F, A]]` instead of an `Eval[F[Cofree[F, A]]]`.

adelbertc commented on free/src/test/scala/cats/free/CofreeTests.scala:
```diff
@@ -0,0 +1,102 @@
+  implicit val iso = CartesianTests.Isomorphisms.invariant[Cofree[Option, ?]]
+
+  checkAll("Cofree[Option, ?]", ComonadTests[Cofree[Option, ?]].comonad[Int, Int, Int])
+  locally {
```
woah what's this `locally` thing do?

kailuowang replied:
http://stackoverflow.com/questions/3237727/what-does-predef-locally-do-and-how-is-it-different-from-predef-identity

edmundnoble commented:
I've added tests for the methods @johnynek pointed out and fixed the tests for `mapBranchingS`/`mapBranchingT`/`mapBranchingRoot`/`unfold` which weren't running (but looked to be).

peterneyens commented on free/src/main/scala/cats/free/Cofree.scala:
```diff
@@ -0,0 +1,135 @@
+
+}
+
+sealed abstract class CofreeInstances2 {
```
Maybe you could make the `CofreeInstancesX` classes and the `CofreeComonad`, `CofreeReducible` and `CofreeTraversable` traits package private (`private[free]`) (mostly so they don't show up in the ScalaDoc) ?

edmundnoble replied:
Very right. One more feature to add to the eventual pedantic-scala tool that will come out once scala.meta has a semantic API ;)

adelbertc commented:
LGTM 👍

xuwei-k commented on free/src/main/scala/cats/free/Cofree.scala:
```diff
@@ -0,0 +1,153 @@
+}
+
+sealed private[free] abstract class CofreeInstances1 extends CofreeInstances2 {
+  /** low priority `Traverse` instance */
```
I wrote the comments in scalaz.Cofree because there are two methos in scalaz.

- https://github.com/scalaz/scalaz/commit/a05ab2c0053d54b37474
- https://github.com/scalaz/scalaz/blob/v7.2.8/core/src/main/scala/scalaz/Cofree.scala#L128

```scala
  /** low priority `Traverse1` instance */
  implicit def cofreeTraverse[F[_]: Traverse]: Traverse1[Cofree[F, ?]] =
```

- https://github.com/scalaz/scalaz/blob/v7.2.8/core/src/main/scala/scalaz/Cofree.scala#L155

```scala
  /** high priority `Traverse1` instance. more efficient */
  implicit def cofreeTraverse1[F[_]: Traverse1]: Traverse1[Cofree[F, ?]] =
```

Cats have only one instance `Traverse[Cofree[F, ?]]`

adelbertc commented:
@xuwei-k Looks like @edmundnoble removed it - what do you think?

tpolecat commented on free/src/main/scala/cats/free/Cofree.scala:
```diff
@@ -0,0 +1,151 @@
+  /** Cofree anamorphism, lazily evaluated. */
+  def unfold[F[_], A](a: A)(f: A => F[A])(implicit F: Functor[F]): Cofree[F, A] =
+    Cofree[F, A](a, Eval.later(F.map(f(a))(unfold(_)(f))))
+
```
It would be nice to have the monadic ana as well.

```scala
def unfoldM[M[_]: Monad, F[_]: Traverse, A](a: A)(f: A => M[F[A]]): M[Cofree[F, A]]
```

edmundnoble replied:
Sure :)

tpolecat commented on free/src/main/scala/cats/free/Cofree.scala:
```diff
@@ -0,0 +1,151 @@
+  /**
+    * A stack-safe algebraic recursive fold out of the cofree comonad.
+    */
+  def cata[F[_], A, B](cof: Cofree[F, A])(folder: (A, F[B]) => Eval[B])(implicit F: Traverse[F]): Eval[B] =
```
I am unhappy that `Eval` forces `Traverse[F]` when the strict version would only need `Functor`.

edmundnoble replied:
Let the record show that I am also made unhappy by this.

edmundnoble commented on free/src/main/scala/cats/free/Cofree.scala:
```diff
@@ -0,0 +1,151 @@
+package free
+
+/**
+  * A free comonad for some branching functor `S`. Branching is done lazily using Eval.
```
Yeppers.

edmundnoble commented on free/src/main/scala/cats/free/Cofree.scala:
```diff
@@ -0,0 +1,151 @@
+  * [[https://github.com/scalaz/scalaz/blob/series/7.3.x/core/src/main/scala/scalaz/Cofree.scala Scalaz's Cofree]],
+  * originally written by Rúnar Bjarnason.
+  */
+final case class Cofree[S[_], A](head: A, tailEval: Eval[S[Cofree[S, A]]]) {
```
The suffix is there to disambiguate it with `tailForced`, but I think it's reasonable for `tail` to be the safe version.

edmundnoble commented:
All cleaned up @tpolecat and @travisbrown, thanks for the reviews :)

tpolecat commented on free/src/main/scala/cats/free/Cofree.scala:
```diff
@@ -0,0 +1,162 @@
+    Cofree[F, A](a, Eval.later(F.map(f(a))(unfold(_)(f))))
+
+  /** Cofree monadic anamorphism, lazily evaluated. */
+  def unfoldM[F[_], M[_], A](a: A)(f: A => M[F[A]])(implicit F: Traverse[F], M: Monad[M]): M[Cofree[F, A]] = {
```
Can you convince me this is stacksafe? I don't quite see it.

tpolecat replied:
(Corrected: I meant to say lazy.)

edmundnoble replied:
I cannot for the life of me figure out how to make it lazy. I have removed it for now just because it's unsafe otherwise, but I'd like to add it back soon.

edmundnoble replied:
:O

dwhitney commented:
unfoldM is correct ;)

On Tue, Dec 20, 2016 at 10:00 PM, Rob Norris <notifications@github.com>
wrote:

> *@tpolecat* commented on this pull request.
> ------------------------------
>
> In free/src/main/scala/cats/free/Cofree.scala
> <https://github.com/typelevel/cats/pull/1446>:
>
> > +    Cofree[S, A](head, Eval.now(tail.value))
> +
> +  /** Evaluate the entire Cofree tree. */
> +  def forceAll(implicit S: Functor[S]): Cofree[S, A] =
> +    Cofree[S, A](head, Eval.now(tail.map(S.map(_)(_.forceAll)).value))
> +
> +}
> +
> +object Cofree extends CofreeInstances {
> +
> +  /** Cofree anamorphism, lazily evaluated. */
> +  def unfold[F[_], A](a: A)(f: A => F[A])(implicit F: Functor[F]): Cofree[F, A] =
> +    Cofree[F, A](a, Eval.later(F.map(f(a))(unfold(_)(f))))
> +
> +  /** Cofree monadic anamorphism, lazily evaluated. */
> +  def unfoldM[F[_], M[_], A](a: A)(f: A => M[F[A]])(implicit F: Traverse[F], M: Monad[M]): M[Cofree[F, A]] = {
>
> FINE
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/typelevel/cats/pull/1446>, or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AACCoYHpgRKN5utktU8H5CLi8jFGlGEEks5rKJZBgaJpZM4Kj4mH>
> .
>

edmundnoble commented:
@dwhitney I tried running it and it looped forever eagerly. It's correct in that it produces the right answer, but it's incorrect in that it makes `Cofree`'s laziness useless.

peterneyens commented:
Thanks @edmundnoble !
